### PR TITLE
os agnostic

### PR DESCRIPTION
**Pull request by richardschneider**
The key path is always OS agnostic and is separated with a `/`. This allows callers to use the standard key name without having to worry about the runtime OS.  

These changes are an improvement on #11 and get all tests to work.

IMHO: #5 is wrong.

Hopefully this will help js-ipfs move forwards on a window platform (https://github.com/ipfs/js-ipfs/issues/1017, https://github.com/ipfs/js-ipfs-repo/issues/144 and https://github.com/ipfs/js-datastore-core/issues/2)

### CODE REVIEW

**Comment by daviddias** [comment]
@richardschneider could you confirm that this feature does not break go-ipfs interop? @dryajov might be able to help :)

**Comment by dryajov** [comment]
I think this PR makes a lot of sense, what I'm not sure about is why the initial `os specific` change was introduced and I haven't had the time to dig into the code to figure it out. Perhaps, @dignifiedquire has some insights as to why the original change was introduced?

@diasdavid I can make sure that interop with Go doesn't break :)

**Comment by richardschneider** [comment]
I'm pretty sure this does not break any go-ipfs interop, I'm using the new code with js-datastore-fs which has a test for interop with go and it passes.

**Comment by Kubuxu** [comment]
EDIT: I read this PR wrong. It removes os specific component from the interface-datastore.

----

The repo should also have interop from Windows to Linux and back. So please check if keys used in for example leveldb are the same on both those systems. For me it seems that platform specific stuff should be done in flatfs and other datastore implementations that touch file system instead of here but I might be wrong.

**Comment by richardschneider** [comment]
@Kubuxu Is this comment directed at me? If so I'm not sure what ur asking.

**Comment by dryajov** [comment]
@Kubuxu This is part of making js-ipfs run on windows. So far I don't think we have any testing going on windows, but we'll keep this in mind - thanks!

@richardschneider I believe, this might potentially require changes to each implementation to account for path conversions for each os? I'll spend some time today looking at the code as well to have a better idea of what else is required.

**Comment by Kubuxu** [comment]
What I am saying is:
datastore can be backed by multiple types of repos: leveldb, badgerds (in Go) and flatfs (and many more). Keys should be exactly the same in all of those repos on all architectures and flatfs (as it currently is the only repo that touches file system so its keys are affected by different separator) should do key translation.

**Comment by Kubuxu** [comment]
The usage of `path.sep` in key "class" might had been invalid "try" for Windows compatibility that isn't needed in this place from the begging.

@whyrusleeping could you take a look at this 

I didn't write this code and I don't understand fully how js's ipfs stack works but something here smells fishy.

**Comment by richardschneider** [comment]
@dryajov The PR is all about making key paths os-agnostic, always using '/'.  I've applied it to `js-datastore-core` and `js-datastore-fs` and they now work on windows.

**Comment by Kubuxu** [comment]
@richardschneider aghr, I always looked at this PR late into the night. You are right, you are removing the usage of `path.sep` here.

**Comment by dryajov** [comment]
@Kubuxu that makes sense. In that case I have a couple of questions:

- Are keys os specific in go?
  - if so, what is the reasoning behind that?

My current understanding of datastore keys is that they _should_ be OS agnostic, and the underlying implementation (fs, leveldb, badger, etc) would convert the hierarchical key to a representation that is specific to its underlying store, i.e. datastore-fs, should take care of making the key separators into some sort of fs structure. (_Take this with a grain of salt, I have not spent much time looking into either the implementations nor the spec, so my reasoning might be a bit of..._)

**Comment by Kubuxu** [comment]
@richardschneider please accept my apology for the mess I caused in this PR. I didn't notice that it was removing the `path.sep` usage here and making things as they should be.

**Comment by dryajov** [comment]
uppps... accidentally closed the PR, ignore.

**Comment by Kubuxu** [comment]
@dryajov keys have the same separator on all platforms in Go `/`.  The flatfs (sharding datastore?) uses system specific function to join paths but it also never receives key with multiple path elements.

What it receives currently is `/CIQEMFC2HN3HPCUI6ZM2BVVUQ6VXLWNJVQZSCW3E6BNL2GT25PMS22I` where the long string is base32 encoded bytes of CID (with no multibase prefix).

**Comment by Kubuxu** [comment]
>  datastore-fs, should take care of making the key separators into some sort of fs structure.

Yes. This PR is correct in this regard.

Side note: in last few days I have misinterpreted quite few PRs for some reason.

**Comment by dryajov** [comment]
@Kubuxu no worries at all, thanks for jumping in 👍

**Comment by richardschneider** [comment]
why is appveyor build not working?  Seems to be some infrastructure issue.  It was working a few days ago, https://ci.appveyor.com/project/dignifiedquire/interface-datastore/build/27

File: package.json [hunk 1]
```diff
@@ -54,7 +54,8 @@
     "David Dias <daviddias.p@gmail.com>",
     "Erin Dachtler <download333@gmail.com>",
     "Juan Batiz-Benet <juan@benet.ai>",
-    "dignifiedquire <dignifiedquire@gmail.com>"
+    "dignifiedquire <dignifiedquire@gmail.com>",
+    "Richard Schneider <makaretu@gmail.com>"
```
Comment (dignifiedquire) [inline package.json:58]: no need to do this, will be picked up from the commits on release

**Comment by dignifiedquire** [comment]
The code looks good to me, but before we merge, lets make sure to have a windows specific version of https://github.com/ipfs/js-ipfs-repo/blob/master/test/interop-test.js using this code to make sure things work as expected

**Comment by richardschneider** [comment]
@dignifiedquire okay will fork and test

**Comment by richardschneider** [comment]
@dignifiedquire tried running tests for `js-ipfs-repo` it fails with an error in `levelup`

```
 1) IPFS Repo Tests on on Node.js default "before all" hook:
: The filename, directory name, or volume label syntax is incorrect.s-ipfs-repo\test\test-repo-for-1509616898591\datastore/MANIFEST-000052

      at c:\Users\Owner\Documents\GitHub\js-ipfs-repo\node_modules\levelup\lib\levelup.js:117:34
      at c:\Users\Owner\Documents\GitHub\js-ipfs-repo\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:39:16
```

Looks like levelup needs some TLC for windows.

**Comment by dignifiedquire** [comment]
looks like the path is not fully normalized and mixes slash types

**Comment by richardschneider** [comment]
I got this error when running from the source in the repo.  My `interface-datastore` code is **NOT** is not being used.  See https://github.com/ipfs/js-ipfs-repo/issues/144

**Comment by dryajov** [comment]
@dignifiedquire I believe that this path normalization should happen at the level of the specific datastore implementation, i.e. https://github.com/ipfs/js-datastore-core/blob/master/src/shard.js in this case?

**Comment by richardschneider** [comment]
@dryajov I tested `js-datastore-core` with this PR and every test passed.

However, the `interface-datastore.tests` against a sharding store were disabled because not all of the query options where implemented in the sharding store.  I've corrected this in PR https://github.com/ipfs/js-datastore-core/pull/3

**Comment by richardschneider** [comment]
@dryajov @diasdavid @dignifiedquire is there anyway you guys can progress PR or ask for changes.  I have a number of the PRs for windows that are waiting on this.  Ideally a new patch release to NPM would be great.

**Comment by dignifiedquire** [comment]
I will see that I give them a round of review tomorrow

**Comment by dryajov** [comment]
@richardschneider let me do a quick review and see if @diasdavid or @dignifiedquire can do a release.

**Review by dryajov — COMMENTED** [review]
LGTM

**Comment by richardschneider** [comment]
@diasdavid or @dignifiedquire anyone want to release?

**Comment by whyrusleeping** [comment]
They are both asleep for a few more hours (european time)

**Comment by daviddias** [comment]
@richardschneider thank you for being awesome and pushing the Windows support throughout multiple modules!

I'm going to start reviewing all of this and to make it easier in all of us, I'm going to merge this PR into a direct branch of this repo. I've invited you to the JavaScript Team of IPFS so that you can contribute to the branch too, Welcome to the party 🎉

**Pull request merged**